### ISSUE TITLE
Precision Error on Payment Amounts

### ISSUE THREAD

**Issue by pkohout-profitsee** [XeroAPI/Xero-Java#132]
Hi,

today I noticed the Problem that some of the Invoices and the according Payments are not summing up on our side. Apparently the conversion from the JSON string to a Float causis this problem. Maybe we should replace the Float with a BigDecimal? See below the JSON that the Xero API sends, and the Java Object that is created (I used the toString() method of Payment.java)

Xeros API response:
```
{
  "PaymentID": "79cda65f-a004-465a-a091-ee8452e5ec24",
  "Date": "\/Date(1551916800000+0000)\/",
  "BankAmount": 148062.76,
  "Amount": 148062.76,
  "Reference": "",
  "CurrencyRate": 1.000000,
  "PaymentType": "ACCPAYPAYMENT",
  "Status": "AUTHORISED",
  "UpdatedDateUTC": "\/Date(1551972182637+0000)\/",
  "HasAccount": true,
  "IsReconciled": false,
  "Account": {
    "AccountID": "ceef66a5-a545-413b-9312-78a53caadbc4",
    "Code": "090"
  },
  "Invoice": {
    "Type": "ACCPAY",
    "InvoiceID": "0a0f6732-7dd0-4287-92aa-465eca07406e",
    "InvoiceNumber": "234",
    "Payments": [],
    "CreditNotes": [],
    "Prepayments": [],
    "Overpayments": [],
    "HasErrors": false,
    "IsDiscounted": false,
    "Contact": {
      "ContactID": "c94a5733-bef8-432f-999d-372f94d10570",
      "ContactNumber": "",
      "Name": "ABC Furniture",
      "Addresses": [],
      "Phones": [],
      "ContactGroups": [],
      "ContactPersons": [],
      "HasValidationErrors": false
    },
    "LineItems": [],
    "CurrencyCode": "USD"
  },
  "HasValidationErrors": false
}
```

And here is the Java Object:
```
class Payment {
    invoice: class Invoice {
        type: ACCPAY
        contact: class Contact {
            contactID: c94a5733-bef8-432f-999d-372f94d10570
            contactNumber: 
            accountNumber: null
            contactStatus: null
            name: ABC Furniture
            firstName: null
            lastName: null
            emailAddress: null
            skypeUserName: null
            contactPersons: []
            bankAccountDetails: null
            taxNumber: null
            accountsReceivableTaxType: null
            accountsPayableTaxType: null
            addresses: []
            phones: []
            isSupplier: null
            isCustomer: null
            defaultCurrency: null
            xeroNetworkKey: null
            salesDefaultAccountCode: null
            purchasesDefaultAccountCode: null
            salesTrackingCategories: null
            purchasesTrackingCategories: null
            trackingCategoryName: null
            trackingCategoryOption: null
            paymentTerms: null
            updatedDateUTC: null
            contactGroups: []
            website: null
            brandingTheme: null
            batchPayments: null
            discount: null
            balances: null
            hasAttachments: null
            validationErrors: null
        }
        lineItems: []
        date: null
        dueDate: null
        lineAmountTypes: null
        invoiceNumber: 234
        reference: null
        brandingThemeID: null
        url: null
        currencyCode: USD
        currencyRate: null
        status: null
        sentToContact: null
        expectedPaymentDate: null
        plannedPaymentDate: null
        subTotal: null
        totalTax: null
        total: null
        totalDiscount: null
        invoiceID: 0a0f6732-7dd0-4287-92aa-465eca07406e
        hasAttachments: null
        payments: []
        prepayments: []
        overpayments: []
        amountDue: null
        amountPaid: null
        fullyPaidOnDate: null
        amountCredited: null
        updatedDateUTC: null
        creditNotes: []
    }
    creditNote: null
    prepayment: null
    overpayment: null
    invoiceNumber: null
    creditNoteNumber: null
    account: class Account {
        code: 090
        name: null
        accountID: ceef66a5-a545-413b-9312-78a53caadbc4
        type: null
        bankAccountNumber: null
        status: null
        description: null
        bankAccountType: null
        currencyCode: null
        taxType: null
        enablePaymentsToAccount: null
        showInExpenseClaims: null
        propertyClass: null
        systemAccount: null
        reportingCode: null
        reportingCodeName: null
        hasAttachments: null
        updatedDateUTC: null
    }
    code: null
    date: 2019-03-07
    currencyRate: 1.0
    amount: 148062.77
    reference: 
    isReconciled: false
    status: AUTHORISED
    paymentType: ACCPAYPAYMENT
    updatedDateUTC: 2019-03-07T15:23:02.637Z
    paymentID: 79cda65f-a004-465a-a091-ee8452e5ec24
    bankAccountNumber: null
    particulars: null
    details: null
}
```

So   `"Amount": 148062.76,` becomes to ` amount: 148062.77`. I also noticed that the BankAmount is send by the API, but is not in the Java Representation.

If you need more information about the Invoice/Invoicepayment setup to reproduce the Problem, just let me know.

Thanks a lot,
Peter

**Comment by SidneyAllen** [comment]
@pkohout-profitsee - what do you think about replacing all floats with double?

**Comment by SidneyAllen** [comment]
@pkohout-profitsee - I've released version 2.3.0 which changes floats to doubles.  My tests indicate it solves the issue you identified.
